### PR TITLE
ARROW-3021: [Go] add support for List arrays

### PR DESCRIPTION
**Pull request by sbinet**
This CL introduces array.Builder, a generic interface to build
array.Interface.
It is needed to be able to build Lists' array values.

array.NewListArray and array.NewListBuilder are introduced to,
respectively, create a new List array value and a builder value to
populate the contents of a List array.

### CODE REVIEW

**Comment by sbinet** [comment]
@stuartcarnie PTAL.

File: go/arrow/array/builder.go [hunk 1]
```diff
@@ -158,7 +181,55 @@ func (b *builder) UnsafeAppendBoolToBitmap(isValid bool) {
...
+	case arrow.STRING:
+	case arrow.BINARY:
+	case arrow.FIXED_SIZE_BINARY:
+	case arrow.DATE32:
+	case arrow.DATE64:
+	case arrow.TIMESTAMP:
+	case arrow.TIME32:
+	case arrow.TIME64:
+	case arrow.INTERVAL:
+	case arrow.DECIMAL:
```
Comment (cristaloleg) [inline go/arrow/array/builder.go]: WDYT about grouping unhandled cases? Ex:
```go
case arrow.STRING, arrow.BINARY, arrow.FIXED_SIZE_BINARY, 
arrow.DATE32, arrow.DATE64, arrow.TIMESTAMP, 
arrow.TIME32, arrow.TIME64, arrow.INTERVAL, 
arrow.DECIMAL:
	panic(fmt.Errorf("arrow/array: unsupported builder for %T", dtype))
```
An same for case above & below 😉

**Review by cristaloleg — APPROVED** [review]

File: go/arrow/array/builder.go (hunk 1 above)
Comment (sbinet) [reply go/arrow/array/builder.go]: leaving it as is would reduce the diff-noise when the support for each of these arrays lands.

but I can change to what you suggest.

**Review by stuartcarnie — APPROVED** [review]
Nice work, @sbinet! Great to see List arrays checked off!

**Pull request closed without merging**